ISO submission editorials (#331)
Green-Software-Foundation/software_carbon_intensity: Software_Carbon_Intensity/Software_Carbon_Intensity_Specification.md

Reducing an SCI score is only possible through the elimination of emissions. That can be achieved by modifying a software system to use less physical hardware, less energy, or consume lower-carbon energy sources. Neutralization or avoidance offsets do not reduce an SCI score ([see exclusions section](#exclusions)). This makes the SCI an ideal strategy that organizations can adopt to meet climate targets focused on eliminating emissions, such as those specified by [1].

The SCI is for everyone. It is possible to calculate an SCI score for any software application, from a large, distributed cloud system to a small monolithic open source library, any on-premise application, or even a serverless function. The environment the product or service is running in can also vary; from personal computers, private data centers or a hyperscale cloud.

> Software practitioners have a significant role to play in collectively reducing the SCI score during the design, development and delivery of software applications. The following list provides some strategies that can be used to do this across different software roles: 

Reviewer: ```suggestion
Software practitioners have a significant role to play in collectively reducing the SCI score during the design, development, and delivery of software applications. The following list provides some strategies that can be used to do this across different software roles: 
```

## Terms and definitions

For the purposes of this document, the following terms and definitions apply.

ISO and IEC maintain terminological databases for use in standardization at the following addresses:
-	ISO Online browsing platform: available at https://www.iso.org/obp
-	IEC Electropedia: available at http://www.electropedia.org/

T.1  
**action**  
explicit outcome taken, or change avoided, depending on the quantifiable emissions measured by this specification

Note to entry: Actions generally relate to using less electricity, using electricity more intelligently, or using less hardware.

T.2  
**carbon-aware**  
attribute of software or hardware that adjusts its behavior (consumption of inputs, processing, or production of outputs) in response to the carbon intensity of the energy it consumes

The following abbreviations are used throughout this specification:
-	E – Energy consumed by a software system

> -	I – Location-based marginal carbon emissions

Reviewer: ```suggestion
-	I – Location-based marginal carbon intensity
```

-	M – Embodied emissions of the hardware needed to operate a software system
-	O – Operational emissions based on the emissions caused by energy consumption

> -	R – Function unit 

Reviewer: ```suggestion
-	R – Functional unit 
```

## Software sustainability actions

All actions that serve to reduce the carbon emissions of a piece of software fit into one of the following categories: 

- **Energy Efficiency**: Actions taken to make software use less electricity to perform the same function.
- **Hardware Efficiency**: Actions taken to make software use fewer physical resources to perform the same function.

> - **Carbon Awareness**: Actions taken to time or region-shift software computation to take advantage of clean, renewable or low carbon sources of electricity.

Reviewer: ```suggestion
- **Carbon Awareness**: Actions taken to time- or region-shift software computation to take advantage of clean, renewable or low carbon sources of electricity.
```

> - **Carbon Awareness**: Actions taken to time or region-shift software computation to take advantage of clean, renewable or low carbon sources of electricity.

Reviewer: ```suggestion
- **Carbon Awareness**: Actions taken to time or region-shift software computation to take advantage of cleaner, more renewable or lower carbon sources of electricity.
```

> - **Carbon Awareness**: Actions taken to time or region-shift software computation to take advantage of clean, renewable or low carbon sources of electricity.

Reviewer: ```suggestion
- **Carbon Awareness**: Actions taken to time- or region-shift software computation to take advantage of cleaner, more renewable or lower carbon sources of electricity.
```

> It is the intent of this specification to encourage more of these actions to be taken during the development and maintenance of software applications.

Reviewer: ```suggestion
It is the intent of this specification to encourage more of these actions to be taken during the design, development, and maintenance of software applications.
```

## Procedure

The steps required to calculate and report an SCI score are:

1. **Bound**: Decide on the [software boundary](#software-boundary); i.e., the components of a software system to include.
1. **Scale**: As the SCI is a rate (carbon emissions per one [functional unit](#functional-unit-r)), pick the functional unit which best describes how the application scales.
1. **Define**: For each software component listed in the software boundary, decide on the [quantification method](#quantification-method); real-world measurements, based on telemetry, or lab-based measurements, based on models.
1. **Quantify**: Calculate a rate for every software component. The SCI value of the whole application is the sum of the SCI values for every software component in the system.
1. **Report**: Disclose the SCI score, software boundary, and the calculation methodology.

## Methodology summary

### General

SCI is a rate; carbon emissions per one unit of `R`. The equation used to calculate the SCI value of a software system is:

`SCI = C per R`

Where:

- The total amount of carbon `C` the software causes to be emitted.

> - All the elements in the SCI equation scale by the same functional unit of `R` (e.g., carbon per additional user, API-call, or ML job).

Reviewer: ```suggestion
- All the elements in the SCI equation scale by the same functional unit of `R` (e.g., carbon emissions per additional user, API-call, or ML training run).
```

This can be expanded to:

`SCI = (O + M) per R`

### Operational emissions

#### General

To calculate the operational emissions associate with software, multiply the electricity consumption of the hardware the software is running on by the regional, granular marginal emissions rate. The marginal emissions rate reflects the change in emissions associated with a change in demand.

> To calculate the operational emissions `O` for a software application, use teh following:

Reviewer: ```suggestion
To calculate the operational emissions `O` for a software application, use the following:
```

`O = (E * I)`

#### Energy

This is the energy consumed by a software system for a functional unit of work. This could be applied to several taxonomies:  

- Datacenter
- Individual machine (e.g., VM/Node)

> - Individual service (e.g., API call or ML training job)

Reviewer: ```suggestion
- Individual service (e.g., API call or ML training run)
```

- Execution of code

Units: this shall be in kilowatt hours (kWh).

#### Location-based marginal carbon intensity

The carbon intensity of electricity is a measure of how much carbon (CO2eq) emissions are produced per kilowatt-hour (kWh) of electricity consumed. Because this specification uses a consequential approach, marginal emissions rates shall be used for electricity consumption. 

Location-based marginal emissions factors measure the grid carbon intensity of a grid region. If the electricity consumption is connected to a grid, the marginal emissions rate of that grid shall be used, which excludes any [market-based measures](#market-based-measures). If the electricity consumption is not connected to a larger regional grid, an appropriate emissions factor for that system shall be used. From a developer perspective, only the location-based info is important in terms of the impact on eliminating carbon emissions. This excludes [market-based measures](#market-based-measures) and is distinct from 100% renewable energy claims.

> The only figure that matters when trying to optimize the scheduling of a computation in real-time is the marginal emissions intensity. This is the emissions intensity of the marginal power plant which will be turned up if a computation is scheduled (e.g., increase electricity demand from the grid) at that moment.

Reviewer: ```suggestion
The only figure that matters when trying to optimize the scheduling of a computation in real-time is the marginal emissions intensity. This is the emissions intensity of the marginal power plant which will need to be turned up if a computation is scheduled (e.g., increase electricity demand from the grid) at that moment.
```

Units: this shall be in grams of carbon per kilowatt hours (gCO2eq/kWh).

### Embodied emissions

Embodied carbon (otherwise referred to as “embedded carbon”) is the amount of carbon emitted during the creation and disposal of a hardware device.

When software runs on a device, a fraction of the total embodied emissions of the device is allocated to the software. This is the value of `M` that needs to be calculated in the SCI equation.

> This fraction consists of both a time-share and a resource-share. The length of time software runs on the device determines its time-share. The percentage of the device reserved just for that application during the time-share determines that application's resource-share.

Reviewer: ```suggestion
This fraction consists of both a time- and resource-share. The length of time software runs on the device determines its time-share. The percentage of the device reserved just for that application during the time-share determines that application's resource-share.
```

> This fraction consists of both a time-share and a resource-share. The length of time software runs on the device determines its time-share. The percentage of the device reserved just for that application during the time-share determines that application's resource-share.

Reviewer: ```suggestion
This fraction consists of both a time-share and a resource-share. The length of time that the software runs on the device determines its time-share. The percentage of the device reserved just for that application during the time-share determines that application's resource-share.
```

> This fraction consists of both a time-share and a resource-share. The length of time software runs on the device determines its time-share. The percentage of the device reserved just for that application during the time-share determines that application's resource-share.

Reviewer: ```suggestion
This fraction consists of both a time- and resource-share. The length of time that the software runs on the device determines its time-share. The percentage of the device reserved just for that application during the time-share determines that application's resource-share.
```

To calculate the share of `M` for a software application, use the equation:

`M = TE * TS * RS`

Where:

- `TE` = Total Embodied Emissions; the sum of Life Cycle Assessment (LCA) emissions for all hardware components.
- `TS` = Time-share; the share of the total life span of the hardware reserved for use by the software.
- `RS` = Resource-share; the share of the total available resources of the hardware reserved for use by the software.

The equation can be expanded further:

`M = TE * (TiR/EL) * (RR/ToR)`

Where:

- `TiR` = Time Reserved; the length of time the hardware is reserved for use by the software.
- `EL` = Expected Lifespan; the anticipated time that the equipment will be installed.
- `RR` = Resources Reserved; the number of resources reserved for use by the software.
- `ToR` = Total Resources; the total number of resources available.

An estimate of all the embodied emissions for the hardware used within the software boundary shall be included.

> Simple models to estimate embodied emissions may be used; however, the most granular data possible and ideally emissions data from a device’s life cycle analysis when calculating the embodied carbon should be used.

Reviewer: ```suggestion
Simple models to estimate embodied emissions may be used; however, the most granular data possible and ideally emissions data from a device’s LCA when calculating the embodied carbon should be used.
```

- resources to support redundancy
- resources to support failover

> If the boundary includes on-premise and/or cloud data center operations, E should take into account the efficiency of the data center, including cooling and other energy consumption necessary to operate a data center. The data center's energy efficiency is usually available as a PUE (Power Usage Effectiveness) value.

Reviewer: ```suggestion
If the boundary includes on-premise and/or cloud data center operations, `E` should take into account the efficiency of the data center, including cooling and other energy consumption necessary to operate a data center. The data center's energy efficiency is usually available as a PUE (Power Usage Effectiveness) value.
```

## Functional unit

> The second step in generating an SCI score is deciding which functional unit will be used to describe how the application scales. First,  decide on the functional unit, using the choice of `R`. Then calculate how much `C` is emitted per unit of `R`.

Reviewer: ```suggestion
The second step in generating an SCI score is deciding which functional unit will be used to describe how the application scales. First, decide on the functional unit, using the choice of `R`. Then calculate how much `C` is emitted per unit of `R`.
```

> For instance, if the application scales by user then choose this as the functional unit.

Reviewer: ```suggestion
For instance, if the application scales by number of users then choose this as the functional unit.
```

### Calculation

What one unit of `R` looks like may be modelled and the total carbon (`C`) calculated for executing one functional unit of work (`R`) in a controlled lab environment. For instance, a benchmark application may be created that models a user interacting with your application and then measure the `C` emitted per run of that benchmark. The result is still a `C` per user.

## Comparing an SCI score to a baseline

When taking an action to reduce the carbon intensity of a piece of software, the intensity should be compared to a baseline. The baseline shall be calculated using an identical methodology to how the proposed SCI was calculated, except excluding the proposed action(s). The measurements, assumptions, models, functional units, etc. shall remain the same between the baseline and proposed SCI.

## Core characteristics

> As this specification matures and develops, the following core characteristics shall remain true:

Reviewer: ```suggestion
As this specification develops, the following core characteristics shall remain true:
```

> - **The SCI is sensitive to carbon awareness, energy efficiency, or hardware efficiency**

Reviewer: ```suggestion
- **The SCI is sensitive to carbon awareness, energy efficiency, and hardware efficiency**
```

>   - The purpose of the SCI is to encourage actions that reduce the carbon emissions of software. Therefore, the SCI shall be sensitive to those actions described in this document under **Software Sustainability Actions**; specifically, carbon awareness, energy efficiency, or hardware efficiency.

Reviewer: ```suggestion
  - The purpose of the SCI is to encourage actions that reduce the carbon emissions of software. Therefore, the SCI shall be sensitive to those actions described in this document under **Software Sustainability Actions**; specifically, carbon awareness, energy efficiency, and hardware efficiency.
```
